<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подключение-библиотек" data-toc-modified-id="Подключение-библиотек-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подключение библиотек</a></span></li><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Предобработка данных</a></span></li><li><span><a href="#Разделение-признаков-и-целевого-признака" data-toc-modified-id="Разделение-признаков-и-целевого-признака-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Разделение признаков и целевого признака</a></span></li><li><span><a href="#Прямое-кодирование-(OHE)" data-toc-modified-id="Прямое-кодирование-(OHE)-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Прямое кодирование (OHE)</a></span></li><li><span><a href="#Разделение-выборок" data-toc-modified-id="Разделение-выборок-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Разделение выборок</a></span></li></ul></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Исследование задачи</a></span><ul class="toc-item"><li><span><a href="#Исследование-баланса-классов" data-toc-modified-id="Исследование-баланса-классов-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Исследование баланса классов</a></span></li><li><span><a href="#Исследование-моделей-без-учёта-дисбаланса" data-toc-modified-id="Исследование-моделей-без-учёта-дисбаланса-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Исследование моделей без учёта дисбаланса</a></span><ul class="toc-item"><li><span><a href="#Решающее-дерево" data-toc-modified-id="Решающее-дерево-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Решающее дерево</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.2.3"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li></ul></li><li><span><a href="#Лучший-вариант-модели" data-toc-modified-id="Лучший-вариант-модели-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Лучший вариант модели</a></span></li></ul></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span><ul class="toc-item"><li><span><a href="#Взвешивание" data-toc-modified-id="Взвешивание-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Взвешивание</a></span></li><li><span><a href="#Увеличение-выборки" data-toc-modified-id="Увеличение-выборки-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Увеличение выборки</a></span></li><li><span><a href="#Уменьшение-выборки" data-toc-modified-id="Уменьшение-выборки-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Уменьшение выборки</a></span></li><li><span><a href="#Лучший-вариант-борьбы" data-toc-modified-id="Лучший-вариант-борьбы-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Лучший вариант борьбы</a></span></li></ul></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Тестирование модели</a></span><ul class="toc-item"><li><span><a href="#Проверка-метрики-f1" data-toc-modified-id="Проверка-метрики-f1-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Проверка метрики f1</a></span></li><li><span><a href="#Исследование-AUC-ROC" data-toc-modified-id="Исследование-AUC-ROC-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Исследование AUC-ROC</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Выводы</a></span><ul class="toc-item"><li><span><a href="#Были-предобработаны-данные-следующим-образом:" data-toc-modified-id="Были-предобработаны-данные-следующим-образом:-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Были предобработаны данные следующим образом:</a></span><ul class="toc-item"><li><span><a href="#Пропуски:" data-toc-modified-id="Пропуски:-6.1.1"><span class="toc-item-num">6.1.1&nbsp;&nbsp;</span>Пропуски:</a></span></li><li><span><a href="#Кодирование:" data-toc-modified-id="Кодирование:-6.1.2"><span class="toc-item-num">6.1.2&nbsp;&nbsp;</span>Кодирование:</a></span></li></ul></li><li><span><a href="#В-ходе-анализа-были-получены-следующие-результаты:" data-toc-modified-id="В-ходе-анализа-были-получены-следующие-результаты:-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>В ходе анализа были получены следующие результаты:</a></span></li><li><span><a href="#Конечный-результат" data-toc-modified-id="Конечный-результат-6.3"><span class="toc-item-num">6.3&nbsp;&nbsp;</span>Конечный результат</a></span></li></ul></li></ul></div>

# Отток клиентов

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Постройте модель с предельно большим значением *F1*-меры. Чтобы сдать проект успешно, нужно довести метрику до 0.59. Проверьте *F1*-меру на тестовой выборке самостоятельно.

Дополнительно измеряйте *AUC-ROC*, сравнивайте её значение с *F1*-мерой.

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

# Проект: "Отток клиентов"

###### Поставленная задача (цель проекта):
- построить модель, которая будет прогнозировать уйдёт клиент из банка в ближайшее время или нет, с предельно большим значением F1-меры (не менее 0.59);
- измерять AUC-ROC и сравнивать её значение с F1-мерой.


###### Имеется информация о (датафрейм):

- клиентах «Бета-Банка»

## Подключение библиотек

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import f1_score, roc_auc_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.utils import shuffle
from sklearn.preprocessing import StandardScaler

In [2]:
pd.set_option('display.max_columns', None)

## Подготовка данных

### Предобработка данных

In [3]:
data = pd.read_csv('Churn.csv')

In [4]:
display(data.head())
print()
data.info()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


<div class="alert alert-secondary" role="alert">
<b>Заметка:</b>
<br> датафрейм состоит из 14 колонок и 10000 строк. 
    
1) Есть пропуски в колонке 'Tenure' - возможно, пропуски стоят у тех клиентов, у которых еще нет полных лет "присутствия" в банке.
    
2) Типы данных, в целом, пойдут
    
    
    В датафрейме представленны следующие колонки:
    
- RowNumber — индекс строки в данных;
- CustomerId — уникальный идентификатор клиента;
- Surname — фамилия;
- CreditScore — кредитный рейтинг;
- Geography — страна проживания;
- Gender — пол;
- Age — возраст;
- Tenure — сколько лет человек является клиентом банка;
- Balance — баланс на счёте;
- NumOfProducts — количество продуктов банка, используемых клиентом;
- HasCrCard — наличие кредитной карты;
- IsActiveMember — активность клиента;
- EstimatedSalary — предполагаемая зарплата;
- Exited — факт ухода клиента.
 
Колонка 'Exited' является целевым признаком, а остальные - просто признаками
</div>

In [5]:
data.duplicated().sum()

0

In [6]:
data['CustomerId'].duplicated().sum()

0

<div class="alert alert-secondary" role="alert">
<b>Заметка:</b>
<br> полных дубликатов и дубликатов по Id нет
    
</div>

In [7]:
data['Gender'].unique()

array(['Female', 'Male'], dtype=object)

<div class="alert alert-secondary" role="alert">
<b>Заметка:</b>
<br> в колонке 'Gender' "лишних" значений нет
    
</div>

In [8]:
data['Age'].describe()

count    10000.000000
mean        38.921800
std         10.487806
min         18.000000
25%         32.000000
50%         37.000000
75%         44.000000
max         92.000000
Name: Age, dtype: float64

<div class="alert alert-secondary" role="alert">
<b>Заметка:</b>
<br> в колонке 'Age' представлены клиенты в возрасте от 18 до 92 лет. Аномальных значений не наблюдается. В основном клиенты банка в возрасте до 45 лет.
</div>


In [9]:
data['Tenure'].describe()

count    9091.000000
mean        4.997690
std         2.894723
min         0.000000
25%         2.000000
50%         5.000000
75%         7.000000
max        10.000000
Name: Tenure, dtype: float64

<div class="alert alert-secondary" role="alert">
<b>Заметка:</b>
<br> клиенты пользуются банком от 0 до 10 лет. В среднем длительность сотрудничества составляет 5 лет. В значениях столбца 'Tenure' присутствует 0, значит, предположение о пропусках для клиентов, кто пользуется банком меньше 1 года, неверно
</div>


In [10]:
data[data['Tenure'].isna()]

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
30           31    15589475    Azikiwe          591     Spain  Female   39   
48           49    15766205        Yin          550   Germany    Male   38   
51           52    15768193  Trevisani          585   Germany    Male   36   
53           54    15702298   Parkhill          655   Germany    Male   41   
60           61    15651280     Hunter          742   Germany    Male   35   
...         ...         ...        ...          ...       ...     ...  ...   
9944       9945    15703923    Cameron          744   Germany    Male   41   
9956       9957    15707861      Nucci          520    France  Female   46   
9964       9965    15642785    Douglas          479    France    Male   34   
9985       9986    15586914     Nepean          659    France    Male   36   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
30       NaN       0.00              3          1               0   
48       NaN  103391.38              1          0               1   
51       NaN  146050.97              2          0               0   
53       NaN  125561.97              1          0               0   
60       NaN  136857.00              1          0               0   
...      ...        ...            ...        ...             ...   
9944     NaN  190409.34              2          1               1   
9956     NaN   85216.61              1          1               0   
9964     NaN  117593.48              2          0               0   
9985     NaN  123841.49              2          1               0   
9999     NaN  130142.79              1          1               0   

      EstimatedSalary  Exited  
30          140469.38       1  
48           90878.13       0  
51           86424.57       0  
53          164040.94       1  
60           84509.57       0  
...               ...     ...  
9944        138361.48       0  
9956        117369.52       1  
9964        113308.29       0  
9985         96833.00       0  
9999         38190.78       0  

[909 rows x 14 columns]

In [11]:
print(f'Количество пропусков в столбце Tenure: {data["Tenure"].isna().sum()}')
print()
print(f'Процент пропусков от общего числа данных: {round(((data["Tenure"].isna().sum())/(data["Tenure"].count()))*100)} %')

Количество пропусков в столбце Tenure: 909

Процент пропусков от общего числа данных: 10 %


In [12]:
data = data.dropna()
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9091 entries, 0 to 9998
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        9091 non-null   int64  
 1   CustomerId       9091 non-null   int64  
 2   Surname          9091 non-null   object 
 3   CreditScore      9091 non-null   int64  
 4   Geography        9091 non-null   object 
 5   Gender           9091 non-null   object 
 6   Age              9091 non-null   int64  
 7   Tenure           9091 non-null   float64
 8   Balance          9091 non-null   float64
 9   NumOfProducts    9091 non-null   int64  
 10  HasCrCard        9091 non-null   int64  
 11  IsActiveMember   9091 non-null   int64  
 12  EstimatedSalary  9091 non-null   float64
 13  Exited           9091 non-null   int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.0+ MB


<div class="alert alert-secondary" role="alert">
<b>Заметка:</b>
<br> пропуски столбца 'Tenure' составляют 10 %, считаю, что их можно удалить
</div>


In [13]:
data['Balance'].describe()

count      9091.000000
mean      76522.740015
std       62329.528576
min           0.000000
25%           0.000000
50%       97318.250000
75%      127561.890000
max      250898.090000
Name: Balance, dtype: float64

<div class="alert alert-secondary" role="alert">
<b>Заметка:</b>
<br> аномальных значений (например, отрицательных) не наблюдается. Пустые счета вполне могут быть :(
</div>


In [14]:
data['NumOfProducts'].describe()

count    9091.000000
mean        1.530195
std         0.581003
min         1.000000
25%         1.000000
50%         1.000000
75%         2.000000
max         4.000000
Name: NumOfProducts, dtype: float64

<div class="alert alert-secondary" role="alert">
<b>Заметка:</b>
<br> аномальных значений не наблюдается. У клиентов, в основном 1 - 2 продукта банка. Максимум - 4
</div>


In [15]:
data['EstimatedSalary'].describe()

count      9091.000000
mean     100181.214924
std       57624.755647
min          11.580000
25%       51227.745000
50%      100240.200000
75%      149567.210000
max      199992.480000
Name: EstimatedSalary, dtype: float64

<div class="alert alert-secondary" role="alert">
<b>Заметка:</b>
<br> в столбце 'EstimatedSalary' есть минимальное значение -  11.58, что выглядит аномально. Возможно, ошиблись в порядке, когда вносили данные в базу. Трогать не буду
</div>


### Разделение признаков и целевого признака

In [16]:
features = data.drop(['RowNumber','CustomerId','Surname','Exited'], axis=1)
target = data['Exited']

display(features.head())
print()
target.head()

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France  Female   42     2.0       0.00              1   
1          608     Spain  Female   41     1.0   83807.86              1   
2          502    France  Female   42     8.0  159660.80              3   
3          699    France  Female   39     1.0       0.00              2   
4          850     Spain  Female   43     2.0  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  
0          1               1        101348.88  
1          0               1        112542.58  
2          1               0        113931.57  
3          0               0         93826.63  
4          1               1         79084.10

0    1
1    0
2    1
3    0
4    0
Name: Exited, dtype: int64

<div class="alert alert-secondary" role="alert">
<b>Заметка:</b>
<br> данные разделены на признаки (столбцы за исключением тех, которые не влияют факт ухода клиента) и целевой признак - 'Exited'
    
</div>

### Прямое кодирование (OHE)

In [17]:
features = pd.get_dummies(features, drop_first=True)
features.head()

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0          619   42     2.0       0.00              1          1   
1          608   41     1.0   83807.86              1          0   
2          502   42     8.0  159660.80              3          1   
3          699   39     1.0       0.00              2          0   
4          850   43     2.0  125510.82              1          1   

   IsActiveMember  EstimatedSalary  Geography_Germany  Geography_Spain  \
0               1        101348.88              False            False   
1               1        112542.58              False             True   
2               0        113931.57              False            False   
3               0         93826.63              False            False   
4               1         79084.10              False             True   

   Gender_Male  
0        False  
1        False  
2        False  
3        False  
4        False

<div class="alert alert-secondary" role="alert">
<b>Заметка:</b>
<br> закодированы столбцы 'Geography' и 'Gender' прямым кодированием
</div>


### Разделение выборок

In [18]:
features_train, features_val_and_test, target_train, target_val_and_test = \
train_test_split(features, target, test_size=0.4, random_state=777)

features_val, features_test, target_val, target_test = \
train_test_split(features_val_and_test, target_val_and_test, test_size=0.5, random_state=777)

In [19]:
display(features_train.head())
print()
display(features_train.info())

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
8284          711   39     3.0  152462.79              1          1   
4133          704   33     3.0       0.00              2          1   
5205          756   28     8.0  179960.20              1          1   
6916          654   24     8.0  145081.73              1          1   
5615          520   29     8.0   95947.76              1          1   

      IsActiveMember  EstimatedSalary  Geography_Germany  Geography_Spain  \
8284               0         90305.97              False            False   
4133               0         73018.74              False             True   
5205               0         89938.08              False            False   
6916               1        130075.07              False            False   
5615               0          4696.44              False            False   

      Gender_Male  
8284        False  
4133         True  
5205         True  
6916        False  
5615        False


<class 'pandas.core.frame.DataFrame'>
Index: 5454 entries, 8284 to 3290
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CreditScore        5454 non-null   int64  
 1   Age                5454 non-null   int64  
 2   Tenure             5454 non-null   float64
 3   Balance            5454 non-null   float64
 4   NumOfProducts      5454 non-null   int64  
 5   HasCrCard          5454 non-null   int64  
 6   IsActiveMember     5454 non-null   int64  
 7   EstimatedSalary    5454 non-null   float64
 8   Geography_Germany  5454 non-null   bool   
 9   Geography_Spain    5454 non-null   bool   
 10  Gender_Male        5454 non-null   bool   
dtypes: bool(3), float64(3), int64(5)
memory usage: 399.5 KB


None

In [20]:
display(features_val.head())
print()
display(features_val.info())

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
7449          698   37     2.0  166178.02              2          1   
3338          688   24     3.0       0.00              2          1   
1133          559   34     4.0       0.00              2          1   
5288          582   30     2.0       0.00              2          1   
65            751   36     6.0  169831.46              2          1   

      IsActiveMember  EstimatedSalary  Geography_Germany  Geography_Spain  \
7449               1         71972.95              False            False   
3338               1        102195.16              False             True   
1133               1         66721.98              False            False   
5288               1        132029.95              False            False   
65                 1         27758.36               True            False   

      Gender_Male  
7449        False  
3338        False  
1133         True  
5288         True  
65          False


<class 'pandas.core.frame.DataFrame'>
Index: 1818 entries, 7449 to 878
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CreditScore        1818 non-null   int64  
 1   Age                1818 non-null   int64  
 2   Tenure             1818 non-null   float64
 3   Balance            1818 non-null   float64
 4   NumOfProducts      1818 non-null   int64  
 5   HasCrCard          1818 non-null   int64  
 6   IsActiveMember     1818 non-null   int64  
 7   EstimatedSalary    1818 non-null   float64
 8   Geography_Germany  1818 non-null   bool   
 9   Geography_Spain    1818 non-null   bool   
 10  Gender_Male        1818 non-null   bool   
dtypes: bool(3), float64(3), int64(5)
memory usage: 133.2 KB


None

In [21]:
display(features_val.head())
print()
display(features_test.info())

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
7449          698   37     2.0  166178.02              2          1   
3338          688   24     3.0       0.00              2          1   
1133          559   34     4.0       0.00              2          1   
5288          582   30     2.0       0.00              2          1   
65            751   36     6.0  169831.46              2          1   

      IsActiveMember  EstimatedSalary  Geography_Germany  Geography_Spain  \
7449               1         71972.95              False            False   
3338               1        102195.16              False             True   
1133               1         66721.98              False            False   
5288               1        132029.95              False            False   
65                 1         27758.36               True            False   

      Gender_Male  
7449        False  
3338        False  
1133         True  
5288         True  
65          False


<class 'pandas.core.frame.DataFrame'>
Index: 1819 entries, 8079 to 2287
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CreditScore        1819 non-null   int64  
 1   Age                1819 non-null   int64  
 2   Tenure             1819 non-null   float64
 3   Balance            1819 non-null   float64
 4   NumOfProducts      1819 non-null   int64  
 5   HasCrCard          1819 non-null   int64  
 6   IsActiveMember     1819 non-null   int64  
 7   EstimatedSalary    1819 non-null   float64
 8   Geography_Germany  1819 non-null   bool   
 9   Geography_Spain    1819 non-null   bool   
 10  Gender_Male        1819 non-null   bool   
dtypes: bool(3), float64(3), int64(5)
memory usage: 133.2 KB


None

<div class="alert alert-secondary" role="alert">
<b>Заметка:</b>
<br> получены три новых датафрейма (тренировочный, валидационный и тестовый) в соотношении 6:2:2. Датафреймы пропусков не содержат, имеют по 11 колонок, 5454 строк в тренировочном датафрейме, 1818 - в валидационном и 1819 - в тестовом
    
</div>

## Исследование задачи

### Исследование баланса классов

In [22]:
print(f'Количество значений положительного класса: {(target == 1).sum()}')
print()
print(f'Количество значений отрицательного класса: {(target == 0).sum()}')

Количество значений положительного класса: 1854

Количество значений отрицательного класса: 7237


<div class="alert alert-secondary" role="alert">
<b>Заметка:</b>
<br> классы не сбалансированы. Их соотношение составляет примерно 2:8
</div>


### Исследование моделей без учёта дисбаланса

#### Решающее дерево

In [23]:
best_depth = 0
best_result_f1_tree = 0


for depth in range(1, 20):
    model = DecisionTreeClassifier(max_depth=depth, random_state=777)
    model.fit(features_train, target_train)
    pred_val = model.predict(features_val)
    result_f1 = f1_score(target_val, pred_val)
    print(f'Максимальная глубина - {depth}, f1 = {result_f1}')
    if best_result_f1_tree < result_f1:
        best_result_f1_tree = result_f1
        best_depth = depth
        
        probabilities_val = model.predict_proba(features_val)
        probabilities_one_val = probabilities_val[:, 1]
        
        auc_roc_tree = roc_auc_score(target_val, probabilities_one_val)
print()      
print(f'Лучший результат f1 = {best_result_f1_tree} при глубине дерева - {best_depth}')
print()
print(f'Значение метрики AUC-ROC: {auc_roc_tree}')

Максимальная глубина - 1, f1 = 0.0
Максимальная глубина - 2, f1 = 0.53125
Максимальная глубина - 3, f1 = 0.416988416988417
Максимальная глубина - 4, f1 = 0.5385878489326765
Максимальная глубина - 5, f1 = 0.5300000000000001
Максимальная глубина - 6, f1 = 0.5548172757475084
Максимальная глубина - 7, f1 = 0.5580645161290322
Максимальная глубина - 8, f1 = 0.556067588325653
Максимальная глубина - 9, f1 = 0.5258358662613981
Максимальная глубина - 10, f1 = 0.530791788856305
Максимальная глубина - 11, f1 = 0.5221238938053098
Максимальная глубина - 12, f1 = 0.5254237288135593
Максимальная глубина - 13, f1 = 0.5232067510548523
Максимальная глубина - 14, f1 = 0.5244565217391304
Максимальная глубина - 15, f1 = 0.5006729475100942
Максимальная глубина - 16, f1 = 0.5013054830287206
Максимальная глубина - 17, f1 = 0.5064935064935066
Максимальная глубина - 18, f1 = 0.49095607235142114
Максимальная глубина - 19, f1 = 0.48774193548387096

Лучший результат f1 = 0.5580645161290322 при глубине дерева - 7

З

<div class="alert alert-secondary" role="alert">
<b>Заметка:</b>
<br> наивысший показатель качества для модели "решающее дерево" получен при гиперпараметре max_depth=7 и составляет 0.558, а auc-roc = 0,819
</div>

#### Случайный лес

In [24]:
best_depth_forest = 0
best_result_f1_forest = 0

for depth in range(1, 30):
    model = RandomForestClassifier(n_estimators=10, max_depth=depth, random_state=777)
    model.fit(features_train, target_train)
    pred_val = model.predict(features_val)
    result_f1 = f1_score(target_val, pred_val)
    print(f'Максимальная глубина - {depth}, f1 = {result_f1}')
    if best_result_f1_forest < result_f1:
        best_result_f1_forest = result_f1
        best_depth_forest = depth

        probabilities_val = model.predict_proba(features_val)
        probabilities_one_val = probabilities_val[:, 1]
        
        auc_roc_forest = roc_auc_score(target_val, probabilities_one_val)
        
        
print()
print(f'Лучший результат f1 = {best_result_f1_forest} при глубине дерева - {best_depth_forest}')
print()
print(f'Значение метрики AUC-ROC: {auc_roc_forest}')

Максимальная глубина - 1, f1 = 0.0
Максимальная глубина - 2, f1 = 0.10268948655256724
Максимальная глубина - 3, f1 = 0.12560386473429952
Максимальная глубина - 4, f1 = 0.2444444444444444
Максимальная глубина - 5, f1 = 0.4604051565377532
Максимальная глубина - 6, f1 = 0.49733570159857904
Максимальная глубина - 7, f1 = 0.5438898450946643
Максимальная глубина - 8, f1 = 0.5819397993311036
Максимальная глубина - 9, f1 = 0.5574324324324325
Максимальная глубина - 10, f1 = 0.584664536741214
Максимальная глубина - 11, f1 = 0.5886178861788619
Максимальная глубина - 12, f1 = 0.5810593900481541
Максимальная глубина - 13, f1 = 0.584664536741214
Максимальная глубина - 14, f1 = 0.5782747603833865
Максимальная глубина - 15, f1 = 0.5736925515055468
Максимальная глубина - 16, f1 = 0.5555555555555556
Максимальная глубина - 17, f1 = 0.5214626391096979
Максимальная глубина - 18, f1 = 0.540453074433657
Максимальная глубина - 19, f1 = 0.5594855305466239
Максимальная глубина - 20, f1 = 0.5525040387722132
Макс

In [25]:
best_est = 0
best_result_f1_forest = 0

for est in range(10, 100, 10):
    model = RandomForestClassifier(n_estimators=est, max_depth=best_depth_forest, random_state=777)
    model.fit(features_train, target_train)
    pred_val = model.predict(features_val)
    result_f1 = f1_score(target_val, pred_val)
    print(f'Количество деревьев - {est}, f1 = {result_f1}')
    if best_result_f1_forest < result_f1:
        best_result_f1_forest = result_f1
        best_est = est
        
        probabilities_val = model.predict_proba(features_val)
        probabilities_one_val = probabilities_val[:, 1]
        
        auc_roc_forest = roc_auc_score(target_val, probabilities_one_val)
        
        
print()
print(f'Лучший результат f1 = {best_result_f1_forest} при количестве деревьев - {best_est}')
print()
print(f'Значение метрики AUC-ROC: {auc_roc_forest}')

Количество деревьев - 10, f1 = 0.5886178861788619
Количество деревьев - 20, f1 = 0.5857605177993528
Количество деревьев - 30, f1 = 0.5825242718446602
Количество деревьев - 40, f1 = 0.5886178861788619
Количество деревьев - 50, f1 = 0.5960912052117264
Количество деревьев - 60, f1 = 0.5888157894736843
Количество деревьев - 70, f1 = 0.5830618892508145
Количество деревьев - 80, f1 = 0.5918699186991869
Количество деревьев - 90, f1 = 0.5931928687196111

Лучший результат f1 = 0.5960912052117264 при количестве деревьев - 50

Значение метрики AUC-ROC: 0.8616934611779973


<div class="alert alert-secondary" role="alert">
<b>Заметка:</b>
<br> наивысший показатель качества для модели "случайный лес" получен при гиперпараметрах max_depth=11 и n_estimators=50, значение f1 - 0.596, а roc-auc=0,86
</div>

#### Логистическая регрессия

In [26]:
best_iter = 0
best_result_f1_regr = 0
for iteration in range(1000, 50000, 5000):
    model = LogisticRegression(max_iter=iteration, random_state=777)
    model.fit(features_train, target_train)
    result_f1 = f1_score(target_val, model.predict(features_val))
    print(f'Количество итераций - {iteration}, f1 = {result_f1}')
    if best_result_f1_regr < result_f1:
        best_result_f1_regr = result_f1
        best_iter = iteration
        
        probabilities_val = model.predict_proba(features_val)
        probabilities_one_val = probabilities_val[:, 1]
        
        auc_roc_regr = roc_auc_score(target_val, probabilities_one_val)
        
        
print()
print(f'Лучший результат f1 = {best_result_f1_regr} при количестве итераций - {best_iter}')
print()
print(f'Значение метрики AUC-ROC: {auc_roc_regr}')

Количество итераций - 1000, f1 = 0.08735632183908046
Количество итераций - 6000, f1 = 0.08735632183908046
Количество итераций - 11000, f1 = 0.08735632183908046
Количество итераций - 16000, f1 = 0.08735632183908046
Количество итераций - 21000, f1 = 0.08735632183908046
Количество итераций - 26000, f1 = 0.08735632183908046
Количество итераций - 31000, f1 = 0.08735632183908046
Количество итераций - 36000, f1 = 0.08735632183908046
Количество итераций - 41000, f1 = 0.08735632183908046
Количество итераций - 46000, f1 = 0.08735632183908046

Лучший результат f1 = 0.08735632183908046 при количестве итераций - 1000

Значение метрики AUC-ROC: 0.6636688054213827


<div class="alert alert-secondary" role="alert">
<b>Заметка:</b>
<br> наивысший показатель качетсва для модели "логистическая регрессия" получен при гиперпараметре max_iter=1000 и составляет 0.087, auc-roc = 0,66
</div>

### Лучший вариант модели

In [27]:
if best_result_f1_tree == max([best_result_f1_tree, best_result_f1_forest, best_result_f1_regr]):
    print(f'Лучшая модель по f1 – решающее дерево с гиперпараметром max_depth = {best_depth}')
elif best_result_f1_forest == max([best_result_f1_tree, best_result_f1_forest, best_result_f1_regr]):
    print(f'Лучшая модель по f1 – случайный лес с гиперпараметром n_estimators = {best_est}')
elif best_result_f1_regr == max([best_result_f1_tree, best_result_f1_forest, best_result_f1_regr]):
    print(f'Лучшая модель по f1 – логистическая регрессия с гиперпараметром max_iter = {best_iter}')

Лучшая модель по f1 – случайный лес с гиперпараметром n_estimators = 50


In [28]:
if auc_roc_tree == max([auc_roc_tree, auc_roc_forest, auc_roc_regr]):
    print(f'Лучшая модель по auc-roc – решающее дерево')
elif auc_roc_forest == max([auc_roc_tree, auc_roc_forest, auc_roc_regr]):
    print(f'Лучшая модель по auc-roc – случайный лес')
elif auc_roc_regr == max([auc_roc_tree, auc_roc_forest, auc_roc_regr]):
    print(f'Лучшая модель по auc-roc – логистическая регрессия')

Лучшая модель по auc-roc – случайный лес


<div class="alert alert-secondary" role="alert">
<b>Заметка:</b>
<br> Лучшая модель – "случайный лес" с гиперпараметрами max_depth=11 и n_estimators=50 как по f1, так и по auc-roc. Значение показателя f1 для данной модели составляет 0.596, auc-roc = 0,86
</div>

## Борьба с дисбалансом

### Взвешивание

In [29]:
model = RandomForestClassifier(random_state=777, class_weight='balanced')
model.fit(features_test, target_test)
pred_val = model.predict(features_val)
f1_balance = f1_score(target_val, pred_val)
print(f'Значение метрики f1: {f1_balance}')


probabilities_val = model.predict_proba(features_val)
probabilities_one_val = probabilities_val[:, 1]
        
auc_roc = roc_auc_score(target_val, probabilities_one_val)

print()
print(f'Значение метрики AUC-ROC: {auc_roc_forest}')

Значение метрики f1: 0.5303292894280762

Значение метрики AUC-ROC: 0.8616934611779973


<div class="alert alert-secondary" role="alert">
<b>Заметка:</b>
<br> взвешивание ухудшило значение f1 на ~0,06, а auc-roc не изменился
</div>


### Увеличение выборки

In [30]:
def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]
    
    features_upcamp = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsamp = pd.concat([target_zeros] + [target_ones] * repeat)
    
    features_upsamp, target_upsamp = shuffle(features_upcamp, target_upsamp, random_state=777)
    
    return features_upsamp, target_upsamp

In [31]:
features_train_upsamp, target_train_upsamp = upsample(features_train, target_train, 4)

print(f'Количество значений положительного класса: {(target_train_upsamp == 1).sum()}')
print()
print(f'Количество значений отрицательного класса: {(target_train_upsamp == 0).sum()}')

Количество значений положительного класса: 4376

Количество значений отрицательного класса: 4360


In [32]:
model = RandomForestClassifier(random_state=777)
model.fit(features_train_upsamp, target_train_upsamp)
pred_val = model.predict(features_val)
f1_upsamp = f1_score(target_val, pred_val)

print(f'Значение метрики f1: {f1_upsamp}')

probabilities_val = model.predict_proba(features_val)
probabilities_one_val = probabilities_val[:, 1]
        
auc_roc = roc_auc_score(target_val, probabilities_one_val)

print()
print(f'Значение метрики AUC-ROC: {auc_roc_forest}')

Значение метрики f1: 0.6115007012622721

Значение метрики AUC-ROC: 0.8616934611779973


<div class="alert alert-secondary" role="alert">
<b>Заметка:</b>
<br> апсэмплинг позволил немного повысить значение f1 (на ~0,02), на auc-roc апсэмплинг не повлиял
</div>

### Уменьшение выборки

In [33]:
def downsample(features, target, fraction):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]
    
    features_downsamp = pd.concat([features_zeros.sample(frac=fraction, random_state=777)] + \
                                  [features_ones])
    target_downsamp = pd.concat([target_zeros.sample(frac=fraction, random_state=777)] + \
                               [target_ones])
    
    features_downsamp, target_downsamp = shuffle(features_downsamp, target_downsamp, random_state=777)
    
    return features_downsamp, target_downsamp

In [34]:
features_train_downsamp, target_train_downsamp = downsample(features_train, target_train, 0.25)

print(f'Количество значений положительного класса: {(target_train_downsamp == 1).sum()}')
print()
print(f'Количество значений отрицательного класса: {(target_train_downsamp == 0).sum()}')

Количество значений положительного класса: 1094

Количество значений отрицательного класса: 1090


In [35]:
model = RandomForestClassifier(random_state=777)
model.fit(features_train_downsamp, target_train_downsamp)
pred_val = model.predict(features_val)
f1_downsamp = f1_score(target_val, pred_val)

print(f'Значение метрики f1: {f1_downsamp}')

probabilities_val = model.predict_proba(features_val)
probabilities_one_val = probabilities_val[:, 1]
        
auc_roc = roc_auc_score(target_val, probabilities_one_val)

print()
print(f'Значение метрики AUC-ROC: {auc_roc_forest}')

Значение метрики f1: 0.58

Значение метрики AUC-ROC: 0.8616934611779973


<div class="alert alert-secondary" role="alert">
<b>Заметка:</b>
<br> даунсэмплинг снизил значение f1 (на ~0,01), auc-roc не изменился
</div>

### Лучший вариант борьбы

In [36]:
if f1_balance == max([f1_balance, f1_upsamp, f1_downsamp]):
    print(f'Побеждает балансирование!')
elif f1_upsamp == max([f1_balance, f1_upsamp, f1_downsamp]):
    print(f'Побеждает увеличение!')
else:
    print(f'Побеждает уменьшение!')

Побеждает увеличение!


<div class="alert alert-secondary" role="alert">
<b>Заметка:</b>
<br> лучшее значение f1 при апсэмплинге, а auc-roc во всех случаях не менялся. Применю его
</div>

## Тестирование модели

### Проверка метрики f1

In [37]:
model = RandomForestClassifier(random_state=777)
model.fit(features_train_upsamp, target_train_upsamp)
pred_test = model.predict(features_test)
f1_test = f1_score(target_test, pred_test)

print(f'Значение метрики f1: {f1_test}')

Значение метрики f1: 0.5946745562130178


In [38]:
if f1_test > 0.59:
    print(f'ТЗ выполнено успешно – метрика f1 выше 0.59')
else:
    print(f'Что-то пошло не так, у меня всё было нормально')

ТЗ выполнено успешно – метрика f1 выше 0.59


<div class="alert alert-secondary" role="alert">
<b>Заметка:</b>
<br> значение f1 на тестовой выборке составляет ~0.595, что превышает заданное значение 0.59
</div>

### Исследование AUC-ROC

In [39]:
probabilities_test = model.predict_proba(features_test)
probabilities_one_test = probabilities_test[:, 1]

auc_roc = roc_auc_score(target_test, probabilities_one_test)

print(f'Значение метрики AUC-ROC: {auc_roc}')

Значение метрики AUC-ROC: 0.8498896493300934


<div class="alert alert-secondary" role="alert">
<b>Заметка:</b>
<br> значение метрики AUC-ROC = 0.85, что весьма близко к идеальному значению - 1. Я вполне довольна. Более того, от auc-roc валидационной выборки отличается не сильно
</div>

## Выводы

### Были предобработаны данные следующим образом:



 #### Пропуски:
- удалены в колонке 'Tenure'

 #### Кодирование:
- закодированы столбцы 'Geography' и 'Gender' прямым кодированием (OHE)


### В ходе анализа были получены следующие результаты:
- получены три новых датафрейма (тренировочный, валидационный и тестовый) в соотношении 6:2:2. Датафреймы пропусков не содержат, имеют по 11 колонок, 5454 строк в тренировочном датафрейме, 1818 - в валидационном и 1819 - в тестовом;
- классы не сбалансированы. Их соотношение составляет примерно 2:8;
- наивысший показатель качества для модели "решающее дерево" получен при гиперпараметре max_depth=7 и составляет 0.558, а auc-roc = 0,819;
- наивысший показатель качества для модели "случайный лес" получен при гиперпараметрах max_depth=11 и n_estimators=50, значение f1 - 0.596, а roc-auc=0,86;
- наивысший показатель качетсва для модели "логистическая регрессия" получен при гиперпараметре max_iter=1000 и составляет 0.087, auc-roc = 0,66;
- лучшая модель – "случайный лес" с гиперпараметрами max_depth=11 и n_estimators=50. Значение показателя качества для данной модели составляет 0.596, а roc-auc=0,86;
- взвешивание ухудшило значение f1 на ~0,06 (с 0,596 до 0,534), а auc-roc не изменился;
- апсэмплинг позволил немного повысить значение f1 до 0,612 (на ~0,02), на auc-roc апсэмплинг не повлиял;
- даунсэмплинг снизил значение f1 до 0,58 (на ~0,01), auc-roc не изменился;
- лучшее значение f1 при апсэмплинге, а auc-roc во всех случаях не менялся;
- значение f1 на тестовой выборке составляет ~0.595, что превышает заданное значение 0.59;
- значение метрики AUC-ROC = 0.85, что весьма близко к идеальному значению - 1 и чуть ниже значения для валидационных данных (ниже на ~0,01)



### Конечный результат
Моделью с максимально большим значением f1 стал «случайный лес» с гиперпараметрами max_depth=11 и n_estimators=50 с применением upsampling. Значение accuracy составило 0.595, что превышает заданное значение - 0.59. Значение метрики AUC-ROC = 0.85